lib/rgeo/cartesian/sweepline_intersector.rb

```diff
@@ -0,0 +1,149 @@
+    # a new line, it will check if it intersects with any of the segments
+    # the line currently intersects at that y value.
+    # This is a more simplistic implementation that uses an array to hold
+    # observed segments instead of a sorted BST, so performance may be significantly
+    # worse in the case of lots of segments overlapping in y-ranges.
+    class SweeplineIntersector
+      Event = Struct.new(:point, :segment, :is_start)
+      Intersection = Struct.new(:point, :s1, :s2)
+
+      def initialize(segments)
+        @segments = segments
+      end
+      attr_reader :segments
+
+      # Returns the "proper" intersections from the list of segments.
+      #
+      # This will only return intersections that are not the start/end or
+      # end/start of the 2 segments. This could be useful for finding intersections
+      # in a ring for example, because knowing that segments are connected in a linestring
+      # is not always helpful, but those are reported by default.
+      #
+      # Note: This is not the true definition of a proper intersection. A
+      # truly proper intersection does not include colinear intersections and
+      # the intersection must lie in the interior of both segments.
+      #
+      # @return [Array<RGeo::Cartesian::SweeplineIntersector::Intersection>]
+      def proper_intersections
+        return @proper_intersections if defined?(@proper_intersections)
+
+        @proper_intersections = []
+        intersections.each do |intersection|
+          s1 = intersection.s1
+          s2 = intersection.s2
+          pt = intersection.point
+
+          unless (pt == s1.s && pt == s2.e) || (pt == s1.e && pt == s2.s)
+            @proper_intersections << intersection
+          end
+        end
+        @proper_intersections
+      end
+
+      # Computes the intersections of the input segments.
+      #
+      # Creates an event queue from the +events+ and adds segments to the
+      # +observed_segments+ array while their ending event has not been popped
+      # from the queue.
+      #
+      # Compares the new segment from the +is_start+ event to each observed segment
+      # then adds it to +observed_segments+. Records any intersections in to the
+      # returned array.
+      #
+      # @return [Array<RGeo::Cartesian::SweeplineIntersector::Intersection>]
+      def intersections
+        return @intersections if defined?(@intersections)
+
+        @intersections = []
+        observed_segments = Set.new
+        events.each do |e|
+          seg = e.segment
+
+          if e.is_start
+            observed_segments.each do |oseg|
+              int_pt = seg.segment_intersection(oseg)
+              if int_pt
+                intersect = Intersection.new(int_pt, seg, oseg)
+                @intersections << intersect
+              end
+            end
+            observed_segments << seg
+          else
+            observed_segments.delete_if { |oseg| oseg == seg }
```

Comment:
Using `#delete_if` runs a `#select` against the keys, making the usage of a set inefficient. `#delete` is of choice here!
```ruby
require "benchmark/ips"
require "set"

set = Array.new(10_000) { rand }.to_set

Benchmark.ips do |x|
  x.report("#delete_if") { r = rand;set.delete_if { _1 == r} }
  x.report("#delete") { r = rand;set.delete(r) }
  x.compare!
end
```

```
Warming up --------------------------------------
          #delete_if   114.000  i/100ms
             #delete   727.950k i/100ms
Calculating -------------------------------------
          #delete_if      1.136k (± 0.7%) i/s -      5.700k in   5.019563s
             #delete      7.224M (± 1.3%) i/s -     36.398M in   5.039573s

Comparison:
             #delete:  7223549.1 i/s
          #delete_if:     1135.6 i/s - 6360.90x  (± 0.00) slower
```

Reply:
Thanks! Good catch.